internal/api/auth/ldap.go

```diff
@@ -1,41 +1,212 @@
+					log.Debug().Msg("LDAP connection is closed, reconnecting")
+					conn, err = l.connect()
+					if err != nil {
+						log.Error().Err(err).Msg("worker LDAP reconnection failed")
+						req.authResp <- false
+						break
+					}
+				}
+
+				// bind with the user credentials
+				var connState connectionStatus
+				ctxTimeout, cancel := context.WithTimeoutCause(ctx, ldap.DefaultTimeout, ErrLDAPTimeout)
+				auth, connState = l.authenticateWithTimeout(ctxTimeout, conn, req.username, req.password)
+				cancel()
+
+				if connState == connectionClosed {
+					log.Debug().Msg("LDAP connection was closed by the server, closing on client side")
+					if err := conn.Close(); err != nil {
+						log.Error().Err(err).Msg("connection was closed by the server but failed to close on client side")
+					}
+				} else {
+					// LDAP connection is still up, we accept the authentication result
+					log.Debug().Msg("auth response valid")
+					break
+				}
+			}
+
+			log.Debug().Msgf("worker LDAP authentication attempt number %d, result: %t", retry, auth)
+
+			req.authResp <- auth
+			tick.Reset(l.maxConnectionLifetime)
+
+		case <-tick.C:
+			// close connection if no request has been made for a minute
+			log.Debug().Msg("timer reached, closing connection")
+			closeLDAPConnection(conn)
+
+		case <-ctx.Done():
+			// gracefully close connection if context is done
+			log.Debug().Msg("context is closed, closing connection")
+			closeLDAPConnection(conn)
+			return
+		}
+	}
+}
+
+func closeLDAPConnection(conn *ldap.Conn) {
+	if conn != nil && !conn.IsClosing() {
+		if err := conn.Close(); err != nil {
+			log.Error().Err(err).Msg("unable to close the LDAP connection")
+		}
+	}
+}
+
+// authenticateWithTimeout performs the authentication against LDAP with a timeout.
+func (l *LDAPAuth) authenticateWithTimeout(ctx context.Context, conn *ldap.Conn, username, password string) (bool, connectionStatus) {
+	// request the authentication
+	res := make(chan result, 1)
+	go func() {
+		a, c := l.authenticate(conn, username, password)
+		res <- result{auth: a, conn: c}
+	}()
+
+	var connState connectionStatus
+	var auth bool
+
+	// handle timeout and context closing
+	select {
+	case r := <-res:
+		auth = r.auth
+		connState = r.conn
+	case <-ctx.Done():
+		auth = false
+		if errors.Is(context.Cause(ctx), ErrLDAPTimeout) {
+			log.Error().Msg("LDAP authentication timeout")
+			closeLDAPConnection(conn)
+		} else {
+			connState = connectionClosed
+		}
+	}
+	return auth, connState
+}
+
+func (l *LDAPAuth) connect() (*ldap.Conn, error) {
+	conn, err := ldap.DialURL(l.ldapURL, ldap.DialWithTLSConfig(l.tlsConfig))
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial to the LDAP server: %w", err)
+	}
+
+	return conn, nil
+}
+
+// authenticate performs the authentication against LDAP.
+// The first returned boolean is true if the authentication is successful, false otherwise.
+// The second returned boolean is true if the connection is closed, false otherwise.
```

Comment:
false if connection is closed, true otherwise

```
const (
	connectionClosed connectionStatus = false
	connectionUp     connectionStatus = true
)
```

Reply:
done